File: protocol-generator/java/src/test/java/org/entitycore/protocol/peer/TypeRegistryTest.java
package org.entitycore.protocol.peer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.junit.jupiter.api.Test;

/**
 * §9.5 type-registry byte-diff — the peer-side dual of the S2 codec corpus, and the
 * golden-file (S8) proof that render-from-model is byte-identical to the oracle's
 * {@code TypeDefinition} entities.
 *
 * <p>Renders all 53 core types (§9.5) from the in-code model ({@link CoreTypeDefs} via
 * {@link CoreTypes#entities()}) and diffs each {@code content_hash} against the canonical
 * cross-impl {@code type-registry-vectors-v1.diag} (the Go-rendered registry). Our hash is
 * 33 bytes (format byte 0x00 ‖ 32-byte SHA-256 digest); the vectors carry
 * {@code ecf-sha256:<64hex>} — we compare the 32-byte digest hex. A green here is what the
 * live oracle's {@code type_system _match} checks confirm independently (53/53).
 */
final class TypeRegistryTest {

    /** name → 64-hex content_hash digest, parsed from the .diag vectors. */
    private static Map<String, String> parseDiag() throws IOException {
        Path diag = Path.of("../shared/test-vectors/v0.8.0/type-registry-vectors-v1.diag");
        List<String> lines = Files.readAllLines(diag);
        Pattern nameP = Pattern.compile("\"name\":\\s*\"([^\"]*)\"");
        Pattern hashP = Pattern.compile("\"content_hash\":\\s*\"ecf-sha256:([0-9a-fA-F]{64})\"");
        java.util.HashMap<String, String> out = new java.util.HashMap<>();
        for (String line : lines) {
            Matcher nm = nameP.matcher(line);
            Matcher hm = hashP.matcher(line);
            if (nm.find() && hm.find()) {
                out.put(nm.group(1), hm.group(1).toLowerCase());
            }
        }
        return out;
    }

    @Test
    void coreFloorRendersByteIdentical() throws IOException {
        Map<String, String> expected = parseDiag();
        int pass = 0;
        StringBuilder fails = new StringBuilder();
        Map<String, Entity> rendered = CoreTypes.entities();
        for (Map.Entry<String, Entity> e : rendered.entrySet()) {
            String name = e.getKey();
            // our hash is 33 bytes: format byte 0x00 ‖ 32-byte digest — compare the digest.
            byte[] full = e.getValue().hash();
            String digestHex = HexFormat.of().formatHex(full, 1, full.length);
            String exp = expected.get(name);
            if (exp == null) {
                fails.append("  ").append(name).append(" — not found in vectors\n");
            } else if (exp.equals(digestHex)) {
                pass++;
            } else {
                fails.append("  ").append(name).append("\n    expected ").append(exp)
                        .append("\n    got      ").append(digestHex).append('\n');
            }
        }
        if (fails.length() > 0) {
            org.junit.jupiter.api.Assertions.fail("type-registry diff failures:\n" + fails);
        }
        assertEquals(53, rendered.size(), "expected exactly the 53-type §9.5 core floor");
        assertEquals(53, pass, "all 53 core types must be byte-identical to the canonical vectors");
    }

    @Test
    void everyCoreTypePublishesScalarSafeEntity() {
        // §1.1: a system/type entity's data is a map; sanity-check the render produces
        // well-formed entities with non-empty content hashes.
        for (Map.Entry<String, Entity> e : CoreTypes.entities().entrySet()) {
            assertNotNull(e.getValue().hash());
            assertEquals(33, e.getValue().hash().length, "ecfv1-sha256 hash is 33 bytes");
        }
    }
}

File: protocol-generator/kotlin/src/test/kotlin/org/entitycore/protocol/peer/TypeRegistryTest.kt
package org.entitycore.protocol.peer

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * §9.5 core type-registry smoke (the type-registry N/N gate).
 *
 * A bootstrapped peer publishes the full §9.5 53-type core floor at
 * `/{peer}/system/type/{name}` (render-from-model — the content_hash of each is computed
 * by THIS peer's S2-green codec over `{type, data}`). This test verifies, for every one
 * of the 53 floor types:
 *  - it is rendered + bound in the store (reachable at its tree path);
 *  - its content_hash is a 33-byte ecfv1-sha256 hash (format byte 0x00 + 32-byte digest);
 *  - the render is DETERMINISTIC (a re-render yields the byte-identical hash) — the
 *    convergence the S4 type_system byte-diff against the canonical vectors depends on.
 *
 * The byte-for-byte diff against the canonical `type-registry-vectors-v1` is the S4
 * `type_system` category; this S3 smoke proves the 53/53 floor renders + binds + is
 * stable, so the registry surface the oracle fetches exists.
 */
class TypeRegistryTest {

    @Test
    fun coreTypeFloorPublishes() {
        val peer = Peer.create(ByteArray(32) { 0x11 })
        val local = peer.localPeer
        val models = CoreTypeDefs.models()
        assertEquals(53, models.size, "the §9.5 core floor is exactly 53 types")

        var checked = 0
        for ((name, _) in models) {
            val e = peer.store.getAt("/$local/system/type/$name")
            assertTrue(e != null, "core type published at tree path: $name")
            assertEquals("system/type", e.type, "$name is a system/type entity")
            val h = e.hash()
            assertEquals(33, h.size, "$name content_hash is 33 bytes (format byte + digest)")
            assertEquals(0, h[0].toInt(), "$name content_hash format byte is 0x00 (ecfv1-sha256)")
            // determinism: a fresh render of the same model yields the byte-identical hash.
            val rerendered = Entity.make("system/type", models.getValue(name))
            assertTrue(rerendered.hash().contentEquals(h), "$name render is deterministic")
            checked++
        }
        println("TYPE-REGISTRY: PASS ($checked/${models.size} core types rendered + bound + stable)")
        assertEquals(53, checked)
    }
}
